pcap/search.go

```diff
@@ -75,21 +64,74 @@ func (s Search) Span() nano.Span {
 	return s.span
 }
 
-func makeMatcher(tcp_ bool, flow Flow) Predicate {
-	match := func(s0, s1 Socket) bool {
+// ID returns an identifier for the search performed.
+func (s Search) ID() string {
+	return s.id
+}
+
+func matchIP(packet gopacket.Packet) (net.IP, net.IP) {
+	network := packet.NetworkLayer()
+	if ip, ok := network.(*layers.IPv4); ok {
+		return ip.SrcIP, ip.DstIP
+	} else if ip, ok := network.(*layers.IPv6); ok {
+		return ip.SrcIP, ip.DstIP
+	}
+	return nil, nil
+}
+
+func genFlowFilter(flow Flow) func(Socket, Socket) bool {
+	return func(s0, s1 Socket) bool {
 		return s0.IP.Equal(flow.S0.IP) && s1.IP.Equal(flow.S1.IP) && s0.Port == flow.S0.Port && s1.Port == flow.S1.Port
 	}
-	return func(tcp bool, src, dst Socket) bool {
-		if tcp != tcp_ {
+}
+
+func genTCPFilter(flow Flow) func(gopacket.Packet) bool {
+	match := genFlowFilter(flow)
+	return func(packet gopacket.Packet) bool {
+		srcIP, dstIP := matchIP(packet)
+		if srcIP == nil {
```

Comment:
I spent a few seconds wondering why there's no `dstIP == nil` check here. Would you consider making matchIP return arguments look like `(src net.IP, dst net.IP, ok bool)`?

Reply:
ok